halo2_proofs/src/poly/multiopen/shplonk.rs

```diff
@@ -58,11 +58,12 @@ fn construct_intermediate_sets<F: FieldExt, I, Q: Query<F>>(queries: I) -> Inter
 where
     I: IntoIterator<Item = Q> + Clone,
 {
+    let queries = queries.into_iter().collect::<Vec<_>>();
+
     // Find evaluation of a commitment at a rotation
     let get_eval = |commitment: Q::Commitment, rotation: Rotation| -> F {
         queries
-            .clone()
-            .into_iter()
+            .iter()
```

Comment:
This might enable the removal of other cloning. I just haven't gone that far.

Reply:
Not sure I follow, which cloning do you mean? I thought this PR already avoids the cloning of the entire chain of queries.

Reply:
You're cloning the entire thing at line 61. 

This would only use `Copy` and avoid any clones at all

Reply:
I see, but it would require some more changes to make the chaining of queries to have `&Query` instead of `Query` in my imagination, so I think collect is fine here.

Reply:
As you wish! 
Anyway, if you say that collecting all them takes 1us, this does not need any further optimization for sure!